src/main/java/homework10/MyHashMap.java

```diff
@@ -0,0 +1,169 @@
+                (!key.equals(positionNode.getKey()) && hash(key) == positionNode.hash) ||
+                !key.equals(positionNode.getKey()) && !(hash(key) == positionNode.hash)) {
+            if (positionNode.next == null) {
+                positionNode.next = new Node<>(key, value, null, hash(key));
+                size++;
+            } else {
+                Node<K, V> nextInBucket = positionNode.next;
+                while (nextInBucket.next != null) {
+                    if (nextInBucket.key.equals(key)) {
+                        V oldValue = nextInBucket.getValue();
+                        nextInBucket.setValue(value);
+                        return oldValue;
+                    }
+                    nextInBucket = nextInBucket.next;
+                }
+                nextInBucket.next = new Node<>(key, value, null, hash(key));
+                size++;
+            }
+        } else if (positionNode.getKey().equals(key)
+                && positionNode.hash == hash(key)) {
+            V oldValue = positionNode.getValue();
+            arrayOfNodes[hash(key) % arrayOfNodes.length].setValue(value);
+            return oldValue;
+        }
+        return null;
+    }
+
+    protected Node<K, V> removeNode(K key) {
+        Node<K, V> currentNode = getNode(key);
+        if (currentNode.next != null) {
+            arrayOfNodes[hash(key) % arrayOfNodes.length] = currentNode.next;
+        } else {
+            arrayOfNodes[hash(key) % arrayOfNodes.length] = null;
+        }
+        size--;
+        return currentNode;
+    }
+
+    protected Node<K, V> getNode(K key) {
+        if (arrayOfNodes[hash(key) % arrayOfNodes.length] != null) {
+            Node<K, V> firstInBucket = arrayOfNodes[hash(key) % arrayOfNodes.length];
+            Node<K, V> nextInBucket = firstInBucket.next;
+            if (firstInBucket.key == null && key == null || firstInBucket.key.equals(key)) {
+                return firstInBucket;
+            }
+            if (nextInBucket != null) {
+                do {
+                    if (nextInBucket.key.equals(key)) {
+                        return nextInBucket;
+                    }
+                } while (nextInBucket.next != null);
+            }
+        }
+        return null;
+    }
+
+    private Node[] resize(Node[] oldArrayOfNodes) {
+        Node[] newArrayOfNodes;
+        if (oldArrayOfNodes.length > 0) {
+            newArrayOfNodes = new Node[oldArrayOfNodes.length * 2];
+        } else {
+            newArrayOfNodes = new Node[DEFAULT_CAPACITY];
+        }
+        for (int i = 0; i < oldArrayOfNodes.length; i++) {
+            Node oldNode;
+            oldNode = oldArrayOfNodes[i];
+            if (oldArrayOfNodes[i] != null) {
+                oldArrayOfNodes[i] = null;
+                if (oldNode.next == null) {
+                    newArrayOfNodes[oldNode.hash % newArrayOfNodes.length] = oldNode;
+                } else {
+                    do {
```

Comment:
better use while cycle

Reply:
if newArrayOfNodes[oldNode.hash % newArrayOfNodes.length] = oldNode will be true on the last iteration, we will lose this result if i'll use while cycle instead of do-while

Reply:
we can rewrite it with while cycle and it will work correctly. next time try doesn't use do-while cycle

Reply:
ok, thx)